src/ai/backend/manager/idle.py

```diff
@@ -368,42 +457,92 @@ async def check_session(self, session: Row, dbconn: SAConnection, policy: Row) -
         # serves as the default fallback if keypair resource policy's idle_timeout is "undefined"
-        idle_timeout = self.idle_timeout.total_seconds()
+        idle_timeout: float = self.idle_timeout.total_seconds()
         # setting idle_timeout:
         # - zero/inf means "infinite"
         # - negative means "undefined"
-        if policy["idle_timeout"] >= 0:
+        if policy["idle_timeout"] > 0:
             idle_timeout = float(policy["idle_timeout"])
-        if (
-            (idle_timeout <= 0)
-            or (math.isinf(idle_timeout) and idle_timeout > 0)
-            or (t - last_access <= idle_timeout)
-        ):
+        if (idle_timeout <= 0) or (math.isinf(idle_timeout) and idle_timeout > 0):
             return True
-        return False
+        idle_time: float = t - last_access
+        remaining: float = idle_timeout - idle_time
+        await redis_helper.execute(
+            redis_obj,
+            lambda r: r.set(
+                self.get_report_key(session_id),
+                msgpack.packb(remaining),
+                ex=int(DEFAULT_CHECK_INTERVAL) * 10,
+            ),
+        )
+        return remaining >= 0
+
+    async def get_checker_result(
+        self,
+        redis_obj: RedisConnectionInfo,
+        report: IdleCheckReport,
+        session_id: SessionId,
+    ) -> IdleCheckReport:
+        key = self.get_report_key(session_id)
+        data = await redis_helper.execute(redis_obj, lambda r: r.get(key))
+        report.timeout = msgpack.unpackb(data) if data is not None else None
+        return report
 
 
 class SessionLifetimeChecker(BaseIdleChecker):
+    terminate_reason: KernelLifecycleEventReason = KernelLifecycleEventReason.IDLE_SESSION_LIFETIME
     name: ClassVar[str] = "session_lifetime"
 
     async def populate_config(self, config: Mapping[str, Any]) -> None:
         pass
 
-    async def check_session(self, session: Row, dbconn: SAConnection, policy: Row) -> bool:
-        now = await dbconn.scalar(sa.select(sa.func.now()))
-        if policy["max_session_lifetime"] > 0:
+    async def report_idle_check(
+        self, kernel: Row, dbconn: SAConnection, policy: Row, redis_obj: RedisConnectionInfo
+    ) -> bool:
+        """
+        Check the kernel has been living longer than resource policy's `max_session_lifetime`.
+        And save remaining time until `max_session_lifetime` of kernel to Redis.
+        """
+
+        result: Optional[float] = None
+        if (max_session_lifetime := policy["max_session_lifetime"]) > 0:
             # TODO: once per-status time tracking is implemented, let's change created_at
             #       to the timestamp when the session entered PREPARING status.
-            if now - session["created_at"] >= timedelta(seconds=policy["max_session_lifetime"]):
-                return False
-        return True
+            idle_timeout = timedelta(seconds=max_session_lifetime)
+            now = await dbconn.scalar(sa.select(sa.func.now()))
```

Comment:
How about using redis' `TIME` command? Most of the components in Backend.AI which requires precise time sync uses that one. Also, in general, using `TIME` command is way faster than querying result of `NOW()` function in postgresql.